### PR TITLE
[WIP][REVIEW-DONE] Apply StateStore's scan operation to the pattern of range scan in stream-stream join v4 

### PR DESCRIPTION
**Pull request by HeartSaVioR**

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/operators/stateful/join/StreamingSymmetricHashJoinExec.scala [hunk 1]
```diff
@@ -893,21 +893,25 @@ case class StreamingSymmetricHashJoinExec(
      */
     def removeOldState(): Long = {
       stateWatermarkPredicate match {
-        case Some(JoinStateKeyWatermarkPredicate(_, stateWatermark)) =>
+        case Some(JoinStateKeyWatermarkPredicate(_, stateWatermark, prevWatermark)) =>
```
Comment (HeartSaVioR) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/operators/stateful/join/StreamingSymmetricHashJoinExec.scala]: Self-review: name isn't 100% clear. `prevStateWatermark`?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/operators/stateful/join/StreamingSymmetricHashJoinExec.scala [hunk 2]
```diff
@@ -893,21 +893,25 @@ case class StreamingSymmetricHashJoinExec(
...
               s.evictByKeyCondition(stateKeyWatermarkPredicateFunc)
 
             case s: SupportsEvictByTimestamp =>
-              s.evictByTimestamp(watermarkMsToStateTimestamp(stateWatermark))
+              s.evictByTimestamp(
+                watermarkMsToStateTimestamp(stateWatermark),
+                prevWatermark.map(watermarkMsToStateTimestamp))
           }
-        case Some(JoinStateValueWatermarkPredicate(_, stateWatermark)) =>
+        case Some(JoinStateValueWatermarkPredicate(_, stateWatermark, prevWatermark)) =>
```
Comment (HeartSaVioR) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/operators/stateful/join/StreamingSymmetricHashJoinExec.scala]: Self-review: same, name isn't 100% clear. prevStateWatermark?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/operators/stateful/join/StreamingSymmetricHashJoinExec.scala [hunk 3]
```diff
@@ -925,21 +929,25 @@ case class StreamingSymmetricHashJoinExec(
      */
     def removeAndReturnOldState(): Iterator[KeyToValuePair] = {
       stateWatermarkPredicate match {
-        case Some(JoinStateKeyWatermarkPredicate(_, stateWatermark)) =>
+        case Some(JoinStateKeyWatermarkPredicate(_, stateWatermark, prevWatermark)) =>
```
Comment (HeartSaVioR) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/operators/stateful/join/StreamingSymmetricHashJoinExec.scala]: Self-review: same, name isn't 100% clear. prevStateWatermark?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/operators/stateful/join/StreamingSymmetricHashJoinExec.scala [hunk 4]
```diff
@@ -925,21 +929,25 @@ case class StreamingSymmetricHashJoinExec(
...
               s.evictAndReturnByKeyCondition(stateKeyWatermarkPredicateFunc)
 
             case s: SupportsEvictByTimestamp =>
-              s.evictAndReturnByTimestamp(watermarkMsToStateTimestamp(stateWatermark))
+              s.evictAndReturnByTimestamp(
+                watermarkMsToStateTimestamp(stateWatermark),
+                prevWatermark.map(watermarkMsToStateTimestamp))
           }
-        case Some(JoinStateValueWatermarkPredicate(_, stateWatermark)) =>
+        case Some(JoinStateValueWatermarkPredicate(_, stateWatermark, prevWatermark)) =>
```
Comment (HeartSaVioR) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/operators/stateful/join/StreamingSymmetricHashJoinExec.scala]: Self-review: same, name isn't 100% clear. prevStateWatermark?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/operators/stateful/join/StreamingSymmetricHashJoinHelper.scala [hunk 5]
```diff
@@ -46,12 +46,18 @@ object StreamingSymmetricHashJoinHelper extends Logging {
     override def toString: String = s"$desc: $expr"
   }
   /** Predicate for watermark on state keys */
-  case class JoinStateKeyWatermarkPredicate(expr: Expression, stateWatermark: Long)
+  case class JoinStateKeyWatermarkPredicate(
+      expr: Expression,
+      stateWatermark: Long,
+      prevBatchStateWatermark: Option[Long] = None)
```
Comment (HeartSaVioR) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/operators/stateful/join/StreamingSymmetricHashJoinHelper.scala]: Self-review: maybe `prevStateWatermark` would be enough?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/operators/stateful/join/StreamingSymmetricHashJoinHelper.scala [hunk 6]
```diff
@@ -46,12 +46,18 @@ object StreamingSymmetricHashJoinHelper extends Logging {
...
     extends JoinStateWatermarkPredicate {
     def desc: String = "key predicate"
   }
   /** Predicate for watermark on state values */
-  case class JoinStateValueWatermarkPredicate(expr: Expression, stateWatermark: Long)
+  case class JoinStateValueWatermarkPredicate(
+      expr: Expression,
+      stateWatermark: Long,
+      prevBatchStateWatermark: Option[Long] = None)
```
Comment (HeartSaVioR) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/operators/stateful/join/StreamingSymmetricHashJoinHelper.scala]: Self-review: maybe `prevStateWatermark` would be enough?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/operators/stateful/join/StreamingSymmetricHashJoinHelper.scala [hunk 7]
```diff
@@ -213,22 +220,28 @@ object StreamingSymmetricHashJoinHelper extends Logging {
           oneSideJoinKeys(joinKeyOrdinalForWatermark.get).nullable)
         val expr = watermarkExpression(Some(keyExprWithWatermark), eventTimeWatermarkForEviction)
         expr.map { e =>
-          // watermarkExpression only provides the expression when eventTimeWatermarkForEviction
```
Comment (HeartSaVioR) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/operators/stateful/join/StreamingSymmetricHashJoinHelper.scala:216]: Self-review: let's not remove the existing comment unless the content is outdated.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/operators/stateful/join/StreamingSymmetricHashJoinHelper.scala [hunk 8]
```diff
@@ -213,22 +220,28 @@ object StreamingSymmetricHashJoinHelper extends Logging {
...
+          StreamingJoinHelper.getStateValueWatermark(
+            attributesToFindStateWatermarkFor = AttributeSet(oneSideInputAttributes),
+            attributesWithEventWatermark = AttributeSet(otherSideInputAttributes),
+            condition,
+            eventTimeWatermarkForLateEvents)
+        }
         val inputAttributeWithWatermark = oneSideInputAttributes.find(_.metadata.contains(delayKey))
         val expr = watermarkExpression(inputAttributeWithWatermark, stateValueWatermark)
         expr.map { e =>
-          // watermarkExpression only provides the expression when eventTimeWatermarkForEviction
```
Comment (HeartSaVioR) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/operators/stateful/join/StreamingSymmetricHashJoinHelper.scala:229]: Self-review: let's not remove the existing comment unless the content is outdated.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/operators/stateful/join/SymmetricHashJoinStateManager.scala [hunk 9]
```diff
@@ -185,14 +186,15 @@ trait SupportsEvictByTimestamp { self: SymmetricHashJoinStateManager =>
   import SymmetricHashJoinStateManager._
 
   /** Evict the state by timestamp. Returns the number of values evicted. */
```
Comment (HeartSaVioR) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/operators/stateful/join/SymmetricHashJoinStateManager.scala]: Self-review: Let's clarify inclusive or exclusive for both parameters.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/operators/stateful/join/SymmetricHashJoinStateManager.scala [hunk 10]
```diff
@@ -185,14 +186,15 @@ trait SupportsEvictByTimestamp { self: SymmetricHashJoinStateManager =>
...
-  def evictByTimestamp(endTimestamp: Long): Long
+  def evictByTimestamp(endTimestamp: Long, startTimestamp: Option[Long] = None): Long
 
   /**
    * Evict the state by timestamp and return the evicted key-value pairs.
```
Comment (HeartSaVioR) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/operators/stateful/join/SymmetricHashJoinStateManager.scala:199]: Self-review: Let's clarify inclusive or exclusive for both parameters.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/operators/stateful/join/SymmetricHashJoinStateManager.scala [hunk 11]
```diff
@@ -647,17 +650,28 @@ class SymmetricHashJoinStateManagerV4(
 
     /**
      * Returns entries where minTs <= timestamp <= maxTs (both inclusive), grouped by timestamp.
-     * Skips entries before minTs and stops iterating past maxTs (timestamps are sorted).
+     * When the range is bounded, uses scanWithMultiValues for efficient range access;
```
Comment (HeartSaVioR) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/operators/stateful/join/SymmetricHashJoinStateManager.scala]: Self-review: need an update of the API name to rangeScanWithMultiValues

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/operators/stateful/join/SymmetricHashJoinStateManager.scala [hunk 12]
```diff
@@ -647,17 +650,28 @@ class SymmetricHashJoinStateManagerV4(
...
+     * falls back to prefixScan for full range to avoid Long overflow on maxTs + 1.
```
Comment (HeartSaVioR) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/operators/stateful/join/SymmetricHashJoinStateManager.scala]: Self-review: ` to avoid Long overflow on maxTs + 1` <- this part isn't probably needed

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/operators/stateful/join/SymmetricHashJoinStateManager.scala [hunk 13]
```diff
@@ -647,17 +650,28 @@ class SymmetricHashJoinStateManagerV4(
...
+      val useScan = minTs != Long.MinValue || maxTs != Long.MaxValue
 
       new NextIterator[GetValuesResult] {
-        private val iter = stateStore.prefixScanWithMultiValues(key, colFamilyName)
+        private val iter = if (useScan) {
+          val startKey = createKeyRow(key, minTs).copy()
+          val endKey = if (maxTs < Long.MaxValue) {
+            Some(createKeyRow(key, maxTs + 1))
+          } else {
+            None
```
Comment (HeartSaVioR) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/operators/stateful/join/SymmetricHashJoinStateManager.scala]: Self-review: Wait, won't this lead to read beyond key? This seems to be dangerous. Shouldn't this case be also prefixScan?

Maybe useScan should be used when `maxTs < Long.MaxValue`? `minTs = Long.MinValue` doesn't seem to matter. Can it bring problem?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/operators/stateful/join/SymmetricHashJoinStateManager.scala [hunk 14]
```diff
@@ -675,18 +689,13 @@ class SymmetricHashJoinStateManagerV4(
...
             val unsafeRowPair = iter.next()
             val ts = TimestampKeyStateEncoder.extractTimestamp(unsafeRowPair.key)
 
-            if (ts > maxTs) {
-              pastUpperBound = true
-              getNext()
-            } else if (ts < minTs) {
-              getNext()
-            } else if (currentTs == -1L || currentTs == ts) {
+            if (currentTs == -1L || currentTs == ts) {
```
Comment (HeartSaVioR) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/operators/stateful/join/SymmetricHashJoinStateManager.scala]: Self-review: this logic is safe when rangeScanWithMultiValues is called, but need detailed review when prefixScanWithMultiValues is called. Can you leave code comment about the safety? Or can you apply filter against the result of prefixScanWithMultiValues to make sure both of prefixScanWithMultiValues and prefixScanWithMultiValues to provide the same result?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/operators/stateful/join/SymmetricHashJoinStateManager.scala [hunk 15]
```diff
@@ -772,9 +781,44 @@ class SymmetricHashJoinStateManagerV4(
...
+    }
+
     // NOTE: This assumes we consume the whole iterator to trigger completion.
-    def scanEvictedKeys(endTimestamp: Long): Iterator[EvictedKeysResult] = {
-      val evictIterator = stateStore.iteratorWithMultiValues(colFamilyName)
+    def scanEvictedKeys(
+        endTimestamp: Long,
+        startTimestamp: Option[Long] = None): Iterator[EvictedKeysResult] = {
+      val startKeyRow = startTimestamp.flatMap { st =>
+        if (st < Long.MaxValue) Some(createScanBoundaryRow(st + 1))
```
Comment (HeartSaVioR) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/operators/stateful/join/SymmetricHashJoinStateManager.scala:852]: Self-review: do we need to build the row with full key schema? Is it because of the spec of state key encoder which expects timestamp to be appended after the last column?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/operators/stateful/join/SymmetricHashJoinStateManager.scala [hunk 16]
```diff
@@ -647,17 +650,28 @@ class SymmetricHashJoinStateManagerV4(
 
     /**
      * Returns entries where minTs <= timestamp <= maxTs (both inclusive), grouped by timestamp.
-     * Skips entries before minTs and stops iterating past maxTs (timestamps are sorted).
+     * When the range is bounded, uses scanWithMultiValues for efficient range access;
+     * falls back to prefixScan for full range to avoid Long overflow on maxTs + 1.
      */
```
Comment (HeartSaVioR) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/operators/stateful/join/SymmetricHashJoinStateManager.scala:670]: Self-review: Can we clarify the inclusive/exclusive for minTs and maxTs? I assume both are inclusive but if it's not, can we leave code comment about it?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/operators/stateful/join/SymmetricHashJoinStateManager.scala [hunk 17]
```diff
@@ -772,9 +781,44 @@ class SymmetricHashJoinStateManagerV4(
...
+      case _ => null
+    }
+
+    private def createScanBoundaryRow(timestamp: Long): UnsafeRow = {
+      val defaultKey = UnsafeProjection.create(keySchema)
+        .apply(defaultInternalRow(keySchema))
+      createKeyRow(defaultKey, timestamp).copy()
+    }
+
     // NOTE: This assumes we consume the whole iterator to trigger completion.
```
Comment (HeartSaVioR) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/operators/stateful/join/SymmetricHashJoinStateManager.scala]: Self-review: Can we clarify the inclusive/exclusive for endTimestamp and startTimestamp? Can we leave code comment about it?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/operators/stateful/join/SymmetricHashJoinStateManager.scala (hunk 15 above)
Comment (HeartSaVioR) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/operators/stateful/join/SymmetricHashJoinStateManager.scala:852]: Self-review: st + 1? Shouldn't it be st? Can you check with inclusive/exclusive on this method and rangeScanWithMultiValues?

Can we have tests which verify with boundary edge cases?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/operators/stateful/join/SymmetricHashJoinStateManager.scala [hunk 18]
```diff
@@ -772,9 +781,44 @@ class SymmetricHashJoinStateManagerV4(
...
-      val evictIterator = stateStore.iteratorWithMultiValues(colFamilyName)
+    def scanEvictedKeys(
+        endTimestamp: Long,
+        startTimestamp: Option[Long] = None): Iterator[EvictedKeysResult] = {
+      val startKeyRow = startTimestamp.flatMap { st =>
+        if (st < Long.MaxValue) Some(createScanBoundaryRow(st + 1))
+        else None
+      }
+      val endKeyRow = if (endTimestamp < Long.MaxValue) {
+        Some(createScanBoundaryRow(endTimestamp + 1))
```
Comment (HeartSaVioR) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/operators/stateful/join/SymmetricHashJoinStateManager.scala:860]: Self-review: I assume this is OK since rangeScanWithMultiValues is exclusive for endKeyRow, but can you double confirm with it?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/operators/stateful/join/SymmetricHashJoinStateManager.scala (hunk 16 above)
Comment (HeartSaVioR) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/operators/stateful/join/SymmetricHashJoinStateManager.scala:670]: Self-review: Can we have tests which verify with boundary edge cases?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/operators/stateful/join/SymmetricHashJoinStateManager.scala [hunk 19]
```diff
@@ -185,13 +185,25 @@ trait SupportsEvictByCondition { self: SymmetricHashJoinStateManager =>
...
-  /** Evict the state by timestamp. Returns the number of values evicted. */
+  /**
+   * Evict the state by timestamp. Returns the number of values evicted.
+   *
+   * @param endTimestamp Inclusive upper bound: evicts entries with timestamp <= endTimestamp.
+   * @param startTimestamp Exclusive lower bound: entries with timestamp <= startTimestamp are
+   *   assumed to have been evicted already (e.g. from the previous batch). When provided,
+   *   the scan starts from startTimestamp + 1.
+   */
   def evictByTimestamp(endTimestamp: Long, startTimestamp: Option[Long] = None): Long
```
Comment (HeartSaVioR) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/operators/stateful/join/SymmetricHashJoinStateManager.scala:196]: Self-review: can callers add +1 and make it be inclusive? It'd be easier to understand. Many cases endTimestamp is exclusive and startTimestamp is inclusive. While we could leave endTimestamp as inclusive, it's probably not good to flip both.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/operators/stateful/join/SymmetricHashJoinStateManager.scala [hunk 20]
```diff
@@ -185,13 +185,25 @@ trait SupportsEvictByCondition { self: SymmetricHashJoinStateManager =>
...
 
   /**
    * Evict the state by timestamp and return the evicted key-value pairs.
    *
    * It is caller's responsibility to consume the whole iterator.
+   *
+   * @param endTimestamp Inclusive upper bound: evicts entries with timestamp <= endTimestamp.
+   * @param startTimestamp Exclusive lower bound: entries with timestamp <= startTimestamp are
```
Comment (HeartSaVioR) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/operators/stateful/join/SymmetricHashJoinStateManager.scala:204]: Self-review: can callers add +1 and make it be inclusive? It'd be easier to understand. Many cases endTimestamp is exclusive and startTimestamp is inclusive. While we could leave endTimestamp as inclusive, it's probably not good to flip both.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/operators/stateful/join/SymmetricHashJoinStateManager.scala [hunk 21]
```diff
@@ -650,22 +662,24 @@ class SymmetricHashJoinStateManagerV4(
...
+     *
+     * When prefixScan is used (maxTs == Long.MaxValue), entries outside [minTs, maxTs] are
+     * filtered out so both code paths produce identical results.
      */
     def getValuesInRange(
         key: UnsafeRow, minTs: Long, maxTs: Long): Iterator[GetValuesResult] = {
       val reusableGetValuesResult = new GetValuesResult()
-      val useScan = minTs != Long.MinValue || maxTs != Long.MaxValue
+      // Only use rangeScan when we have a bounded endKey (maxTs < Long.MaxValue);
+      // without an upper bound the scan could read beyond this key's entries.
```
Comment (HeartSaVioR) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/operators/stateful/join/SymmetricHashJoinStateManager.scala]: Self-review: maybe the comment is for the old code. Now the reason we disallow maxTs == Long.MaxValue is just because we can't apply +1 to make it be exclusive. Let's update the code comment.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/operators/stateful/join/SymmetricHashJoinStateManager.scala [hunk 22]
```diff
@@ -799,20 +818,39 @@ class SymmetricHashJoinStateManagerV4(
...
+     * Build a scan boundary row for rangeScan. The TsWithKeyTypeStore uses
+     * TimestampAsPrefixKeyStateEncoder, which encodes the row as [timestamp][key_fields].
+     * We need a full-schema row (not just the timestamp) because the encoder expects all
+     * key columns to be present. Default values are used for the key fields since only the
+     * timestamp matters for ordering in the prefix encoder.
+     */
     private def createScanBoundaryRow(timestamp: Long): UnsafeRow = {
       val defaultKey = UnsafeProjection.create(keySchema)
         .apply(defaultInternalRow(keySchema))
       createKeyRow(defaultKey, timestamp).copy()
```
Comment (HeartSaVioR) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/operators/stateful/join/SymmetricHashJoinStateManager.scala:833]: Self-review: shall we check how many times we skip copy() from createKeyRow()? Looks like many callers call copy() anyway, which we could do it in createKeyRow() if there is no clear benefit, since there are multiple callers and reasoning about the safety of UnsafeRow from createKeyRow is getting harder.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/operators/stateful/join/SymmetricHashJoinStateManager.scala [hunk 23]
```diff
@@ -799,20 +818,39 @@ class SymmetricHashJoinStateManagerV4(
...
 
-    // NOTE: This assumes we consume the whole iterator to trigger completion.
+    /**
+     * Scan keys eligible for eviction within the timestamp range.
+     *
+     * This assumes we consume the whole iterator to trigger completion.
+     *
+     * @param endTimestamp Inclusive upper bound: entries with timestamp <= endTimestamp are
+     *   eligible for eviction.
+     * @param startTimestamp Exclusive lower bound: entries with timestamp <= startTimestamp
```
Comment (HeartSaVioR) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/operators/stateful/join/SymmetricHashJoinStateManager.scala:843]: Self-review: can callers add +1 and make it be inclusive? It'd be easier to understand. Many cases endTimestamp is exclusive and startTimestamp is inclusive. While we could leave endTimestamp as inclusive, it's probably not good to flip both.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/operators/stateful/join/SymmetricHashJoinStateManager.scala
Comment (HeartSaVioR) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/operators/stateful/join/SymmetricHashJoinStateManager.scala:862]: Self-review: maybe good to have a code comment about safety of None here?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/SymmetricHashJoinStateManagerSuite.scala [hunk 24]
```diff
@@ -1105,6 +1105,96 @@ class SymmetricHashJoinStateManagerEventTimeInValueSuite
     }
   }
 
+  test("StreamingJoinStateManager V4 - getValuesInRange boundary edge cases") {
```
Comment (HeartSaVioR) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/SymmetricHashJoinStateManagerSuite.scala:1108]: Self-review: shall we update the test while we flip inclusive <-> exclusive from the prior review comment?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/SymmetricHashJoinStateManagerSuite.scala [hunk 25]
```diff
@@ -1105,6 +1105,96 @@ class SymmetricHashJoinStateManagerEventTimeInValueSuite
...
+      // Range entirely outside stored timestamps
+      assert(getJoinedRowTimestamps(40, Some((100L, 200L))) === Seq.empty)
+      assert(getJoinedRowTimestamps(40, Some((1L, 5L))) === Seq.empty)
+
+      // Full range via None (all entries)
+      assert(getJoinedRowTimestamps(40, None) === Seq(10, 20, 30, 40, 50))
+    }
+  }
+
+  test("StreamingJoinStateManager V4 - evictByTimestamp with startTimestamp boundary cases") {
```
Comment (HeartSaVioR) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/SymmetricHashJoinStateManagerSuite.scala]: Self-review: shall we update the test while we flip inclusive <-> exclusive from the prior review comment?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/SymmetricHashJoinStateManagerSuite.scala [hunk 26]
```diff
@@ -1105,6 +1105,96 @@ class SymmetricHashJoinStateManagerEventTimeInValueSuite
...
+      val evictByTs = manager.asInstanceOf[SupportsEvictByTimestamp]
+      val evictedValues = evictByTs.evictAndReturnByTimestamp(30, Some(10))
+        .map(p => toValueInt(p.value)).toSeq.sorted
+      // startTimestamp=10 is exclusive, endTimestamp=30 is inclusive: timestamps 20 and 30
+      assert(evictedValues === Seq(20, 30))
+      assert(get(40) === Seq(10, 40, 50))
+    }
+  }
+
+  test("StreamingJoinStateManager V4 - evictByTimestamp boundary: start equals end") {
```
Comment (HeartSaVioR) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/SymmetricHashJoinStateManagerSuite.scala]: Self-review: shall we update the test while we flip inclusive <-> exclusive from the prior review comment?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/SymmetricHashJoinStateManagerSuite.scala [hunk 27]
```diff
@@ -1105,6 +1105,96 @@ class SymmetricHashJoinStateManagerEventTimeInValueSuite
...
+
+      // startTimestamp=20 (exclusive) and endTimestamp=20 (inclusive): range is empty
+      val evictByTs = manager.asInstanceOf[SupportsEvictByTimestamp]
+      val removed = evictByTs.evictByTimestamp(20, Some(20))
+      assert(removed === 0)
+      assert(get(40) === Seq(10, 20, 30))
+    }
+  }
+
+  test("StreamingJoinStateManager V4 - evictByTimestamp boundary: start just below entry") {
```
Comment (HeartSaVioR) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/SymmetricHashJoinStateManagerSuite.scala]: Self-review: shall we update the test while we flip inclusive <-> exclusive from the prior review comment?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/SymmetricHashJoinStateManagerSuite.scala (hunk 24 above)
Comment (HeartSaVioR) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/SymmetricHashJoinStateManagerSuite.scala:1108]: Self-review: we do have so many test cases for boundary edge case now; if they can share the input data, shall we consolidate them into one? Maybe ideally just two test cases, 1) getValuesInRange 2) evictByTimestamp.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/operators/stateful/join/SymmetricHashJoinStateManager.scala (hunk 19 above)
Comment (HeartSaVioR) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/operators/stateful/join/SymmetricHashJoinStateManager.scala:196]: I changed my mind. Should be OK.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/operators/stateful/join/SymmetricHashJoinStateManager.scala (hunk 20 above)
Comment (HeartSaVioR) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/operators/stateful/join/SymmetricHashJoinStateManager.scala:204]: I changed my mind. Should be OK.

**Comment by HeartSaVioR** [comment]
https://github.com/apache/spark/pull/55267
